src/utils.ts

```diff
@@ -63,4 +65,112 @@ export namespace Utils {
+   * challenge transaction.
+   *
+   * This function performs the following checks:
+   *
+   *   1. Verifies that the transaction's source is the same as the server account id.
+   *   2. Verifies that the number of operations in the transaction is equal to one and of type manageData.
+   *   3. Verifies if timeBounds are still valid.
+   *   4. Verifies if the transaction has been signed by the server and the client.
+   *   5. Verifies that the sequenceNumber is equal to zero.
+   *
+   *
+   * @see [SEP0010: Stellar Web Authentication](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0010.md)
+   * @function
+   * @memberof Utils
+   * @param {string} challengeTx SEP0010 transaction challenge transaction in base64.
+   * @param {string} serverAccountID The server's stellar account.
+   * @example
+   * import { Utils, Network }  from 'stellar-sdk'
+   *
+   * Network.useTestNetwork();
+   *
+   * let challenge = Utils.verifyChallengeTx("base64tx", "server-account-id")
+   * @returns {boolean}
+   */
+  export function verifyChallengeTx(
+    challengeTx: string,
+    serverAccountId: string,
+  ): boolean {
+    const transaction = new Transaction(challengeTx);
+
+    const sequence = Number.parseInt(transaction.sequence, 10);
+
+    if (sequence !== 0) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction sequence number should be zero",
+      );
+    }
+
+    if (transaction.source !== serverAccountId) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction source account is not equal to the server's account",
+      );
+    }
+
+    if (transaction.operations.length !== 1) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction should contain only one operation",
+      );
+    }
+
+    const [operation] = transaction.operations;
+
+    if (!operation.source) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction's operation should contain a source account",
+      );
+    }
+
+    if (operation.type !== "manageData") {
+      throw new InvalidSep10ChallengeError(
+        "The transaction's operation should be manageData",
+      );
+    }
+
+    if (!signedBy(transaction, serverAccountId)) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction is not signed by the server",
+      );
+    }
+
+    if (!signedBy(transaction, operation.source as string)) {
+      throw new InvalidSep10ChallengeError(
+        "The transaction is not signed by the client",
+      );
+    }
+
+    if (!validateTimebounds(transaction)) {
+      throw new InvalidSep10ChallengeError("The transaction has expired");
+    }
+
+    return true;
+  }
+
+  function validateTimebounds(transaction: Transaction): boolean {
+    if (!transaction.timeBounds) {
+      return false;
+    }
+
+    const now = Math.floor(Date.now() / 1000);
```

Comment:
I'm not sure about validating timebounds locally. Ideally we would validate against the server's time.

Reply:
@morleyzhi good call, I think we could go with:

1. Let the server pass the current time in epoch and we use that to compare.
2. Not include timebounds validation and leave that to the implementer.

I'm biased towards 1, thoughts?

Reply:
(1) sounds right. I'm not sure if this is what you mean by "let the server pass the current time", but fetchTimebounds gets the server time from HorizonAxiosClient, which keeps track of the timestamps it gets from Horizon calls:

https://github.com/stellar/js-stellar-sdk/blob/0be789888812e31dc05ae83b7e8d3db1c7318b2b/src/server.ts#L124

Reply:
@morleyzhi After thinking about this, I think it is okay to leave the timebound check to be local, since this is not related with Horizon but rather helping the server doing auth via SEP10 to check that it is valid.

This code will be run on the server which originally generated the SEP10 challenge, which is the source of truth for this time-bounds.